### ISSUE TITLE
solve_quartic returns wrong real roots for ~28% of general quartics: the y-selection argmax ranks solve_cubic's zero placeholders, and the R≈0 fallback uses an absolute tolerance on sqrt(R_sq)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4346]
`solve_quartic` returns wrong real roots for a large share of ordinary inputs. Two independent defects, both in the Ferrari step after the resolvent cubic; neither is touched by #4338 or #4339, which only guard gradients, and neither is seen by the test suite because `TestQuarticSolver::test_random` builds every polynomial from four *real* roots.

Measured on `main` `217b92e6`, torch 2.14.0, CPU, float64 unless stated, against `numpy.roots` (a returned entry counts as a real root only if some numpy real root is within 1e-5 of it; "spurious" is a nonzero entry that matches no real root):

| coefficients (leading 1, others uniform in [-5, 5]) | rows | rows with a missing real root | spurious entries |
| --- | --- | --- | --- |
| general | 2000 | 558 (28 %) | 1750 |
| `x⁴ + b x³ + e` | 2000 | 1003 (50 %) | 1003 |
| `x⁴ + b x³ + e`, `e < 0` | 2000 | 2000 (100 %) | 2000 |
| biquadratic `x⁴ + c x² + e` | 2000 | ~55 % | 1042 |
| `x⁴ + d x + e` with `d = ±1e-6` | 2000 | 1886 (94 %), worst error 5.6e-3 | 1886 |
| `x⁴ + d x + e`, `d` uniform | 2000 | 0 | 0 |

Clean repros (the first two behave the same in float32; the 1e-6 ones need float64, float32's `zero_tol=1e-6` happens to catch them):

```python
import torch
from kornia.geometry.solvers import solve_quartic
solve_quartic(torch.tensor([[1., 2., 0., 0., -16.]], dtype=torch.float64))
# tensor([[ 0., -0.,  0., -2.]])      real roots are -2.760555 and 1.638345; -2 is not a root (16-16-16 = -16)
solve_quartic(torch.tensor([[1., 0., -4., 0., -5.]], dtype=torch.float64))
# tensor([[ 2.,  0.,  0., -2.]])      x⁴ - 4x² - 5 = (x²-5)(x²+1), roots ±2.236068; ±2 are not roots
solve_quartic(torch.tensor([[1., 0., 0., 1e-6, -16.]], dtype=torch.float64))
# tensor([[0., 0., 1.988574, -1.988574]])   roots are ±2 to 1e-7; a 1e-6 perturbation moved them by 1.1e-2
solve_quartic(torch.tensor([[1., 1e-6, 0., 0., -16.]], dtype=torch.float64))
# tensor([[0., -0., 0., -1e-6]])      the ±2 roots of x⁴ - 16 vanish for any b in [1e-6, 0.1]
```

## Defect 1 — the "pick y maximising R²" step ranks `solve_cubic`'s zero placeholders as candidates

`solve_cubic` returns `0` in the slots of non-real roots (documented). `solve_quartic` then does

```python
R_sq_candidates = 0.25 * A_sq.unsqueeze(-1) - B.unsqueeze(-1) + y_roots
best_idx = torch.argmax(R_sq_candidates, dim=-1, keepdim=True)
```

over all three slots, so a placeholder `y = 0` competes with the real resolvent root. Whenever the resolvent has one real root and it is negative, the placeholder wins (`A²/4 - B + 0 > A²/4 - B + y`), Ferrari continues with `y = 0`, and both quadratics are wrong. For `x⁴ + 2x³ - 16` the resolvent is `y³ + 64y + 64`, `solve_cubic` gives `[-0.985, 0, 0]`, the candidates are `[0.015, 1, 1]`, and `argmax` picks a placeholder.

Every one of the 558 general failures above has exactly this shape (checked row by row). Masking the candidates by "is a root of the resolvent" (`|p(y)| ≤ tol`, `-inf` otherwise) before the `argmax` takes the general case to **0 missing / 0 spurious** on the same 2000 rows. `solve_cubic` returning a real-root mask alongside the roots, or the quartic recomputing it from the residual, are both fine; the argmax must not see placeholders.

## Defect 2 — the `R ≈ 0` fallback keys on an absolute `zero_tol` of `R`, so rounding noise in `R_sq` sends `R ≈ 1e-8` down the division path

```python
R = torch.sqrt(torch.clamp(R_sq, min=0.0))
mask_R_small = torch.abs(R) < zero_tol          # 1e-6 float32, 1e-12 float64
E[mask_R_large] = (A*y - 2*C) / (4*R)
```

For a biquadratic (or a near-biquadratic) `R_sq` is analytically 0 but comes out of the cubic as `±1e-16`; `sqrt` turns `1e-16` into `1e-8`, which is far above `1e-12`, and `E` is computed as `(≈0)/(4e-8)` — arbitrary. That is the ±1.9886 result for `x⁴ + 1e-6 x - 16` and the ~55 % biquadratic failure rate that survives the defect-1 fix (588 / 2000 still missing with placeholders masked). The threshold has to be relative to the coefficient scale and applied to `R_sq`, not to `R` after a `sqrt`.

## Test gap

`test_random` constructs `(x-r1)(x-r2)(x-r3)(x-r4)` from `randn` roots, so every polynomial has four real roots and a resolvent with three real roots; neither branch above is ever exercised. The pin for this issue should draw coefficients directly (or two real roots plus a complex pair) and compare with `numpy.roots` real parts, plus the four literals above.

## Scope

No in-tree caller (`git grep solve_quartic kornia/` hits only the exports), so this is the public API only. Found while reviewing #4339's biquadratic pins: the finite-difference derivative at `x⁴ - 16` could not be measured because the forward is discontinuous there. #4338 (batch leak in `solve_cubic`, #4334) and #4339 (`sqrt` guards) are gradient fixes and stay valid; whoever takes this should land it after them to avoid three-way conflicts in `polynomial_solver.py` and `TestQuarticSolver`.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).
